presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ParametricScalarImplementation.java

```diff
@@ -344,17 +407,17 @@ else if (parameterType.isPrimitive() && !hasNullFlag) {
                         checkArgument(!nullableArgument, "Method [%s] has parameter with primitive type %s annotated with @SqlNullable", method, parameterType.getSimpleName());
                     }
 
-                    if (typeParameterNames.contains(type.value()) && !(parameterType == Object.class && nullableArgument)) {
-                        // Infer specialization on this type parameter. We don't do this for @SqlNullable Object because it could match a type like BIGINT
-                        Class<?> specialization = specializedTypeParameters.get(type.value());
-                        Class<?> nativeParameterType = Primitives.unwrap(parameterType);
-                        checkArgument(specialization == null || specialization.equals(nativeParameterType), "Method [%s] type %s has conflicting specializations %s and %s", method, type.value(), specialization, nativeParameterType);
-                        specializedTypeParameters.put(type.value(), nativeParameterType);
+                    inferSpecialization(method, parameterType, type, nullableArgument);
```

Comment:
With these 3 boolean flags `nullableArgument`, `hasNullFlag`, and `isBlockAndPositionNullConvention`, there exists 8 possible combinations. However, only 4 of them is legal.

Previously, there were 4 possible combinations, but only 3 of them is legal. That was not ideal. However, It was not too hard to 1) conclude that the illegal one is impossible; 2) parser fails loudly when both `@IsNull` and `@SqlNullable` is present.

With the addition, reasoning about that just became pretty much intractable.

There are two alternatives ways to address this:

1. This code is organized in such a way that consumption of these variables and production of these variables are clearly split. At the split, add assertion.
2. Use an enum instead of 3 flags.

I highly recommend taking the 2nd alternative.